src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -41,55 +43,107 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_copy_ep_t, uct_ep_t);
     ucs_trace_data("%s [ptr %p len %zu] to 0x%" PRIx64, _name, (_iov)->buffer, \
                    (_iov)->length, (_remote_addr))
 
-#define UCT_CUDA_COPY_CHECK_AND_CREATE_STREAM(_iface, _id) \
-    if ((_iface)->stream[_id] == 0) { \
-        ucs_status_t __status; \
-        __status = UCT_CUDAR_CALL_LOG_ERR(cudaStreamCreateWithFlags, \
-                                          &(_iface)->stream[_id], \
-                                          cudaStreamNonBlocking); \
-        if (UCS_OK != __status) { \
-            return UCS_ERR_IO_ERROR; \
-        } \
+static UCS_F_ALWAYS_INLINE cudaStream_t *
+uct_cuda_copy_get_stream(uct_cuda_copy_iface_t *iface,
+                         ucs_memory_type_t src_type,
+                         ucs_memory_type_t dst_type)
+{
+    cudaStream_t *stream;
+    ucs_status_t status;
+
+    if ((src_type == UCS_MEMORY_TYPE_LAST) &&
+        (dst_type == UCS_MEMORY_TYPE_LAST)) {
+        stream = &iface->short_stream;
+    } else if ((src_type < UCS_MEMORY_TYPE_LAST) &&
+               (dst_type < UCS_MEMORY_TYPE_LAST)) {
+        stream = &iface->queue_desc[src_type][dst_type].stream;
+    } else {
+        goto err;
+    }
+
+    if (*stream == 0) {
+        status = UCT_CUDA_FUNC_LOG_ERR(cudaStreamCreateWithFlags(stream,
+                                                                 cudaStreamNonBlocking));
+        if (status != UCS_OK ) {
+            goto err;
+        }
+    }
+
+    return stream;
+
+err:
+    return NULL;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_memory_type_t
+uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
+{
+    ucs_memory_info_t mem_info;
+    ucs_status_t status;
+
+    status = ucs_memtype_cache_lookup(address, length, &mem_info);
+    if ((status == UCS_ERR_NO_ELEM) ||
+        ((mem_info.type == UCS_MEMORY_TYPE_UNKNOWN))) {
+        status = uct_cuda_base_detect_memory_type(md, address, length,
```

Comment:
should we update the cache?

Reply:
`uct_cuda_base_detect_memory_type` updates the cache.